src/main/haskell/language-kore/src/Data/Kore/AST/MetaOrObject.hs

```diff
@@ -1,64 +1,107 @@
-    isMeta = not . isObject
-    {-# MINIMAL isObject | isMeta #-}
+    isMeta l = case isMetaOrObject l of IsMeta -> True; _ -> False
+    {-# MINIMAL isMetaOrObject #-}
 
 instance MetaOrObject Meta where
-    isMeta _ = True
+    isMetaOrObject _ = IsMeta
 instance MetaOrObject Object where
-    isObject _ = True
-
-data MetaOrObjectTransformer thing result = MetaOrObjectTransformer
-    { metaTransformer   :: thing Meta -> result
-    , objectTransformer :: thing Object -> result
-    }
-
-applyMetaObjectFunction
-    :: (Typeable thing, MetaOrObject level)
-    => thing level -> MetaOrObjectTransformer thing c -> c
-applyMetaObjectFunction x = applyMetaObjectFunctionCasted (cast x) (cast x)
-applyMetaObjectFunctionCasted
-    :: Maybe (thing Object)
-    -> Maybe (thing Meta)
-    -> MetaOrObjectTransformer thing c
-    -> c
-applyMetaObjectFunctionCasted (Just x) Nothing f = objectTransformer f x
-applyMetaObjectFunctionCasted Nothing (Just x) f = metaTransformer f x
-applyMetaObjectFunctionCasted _ _ _ =
-    error "applyMetaObjectFunctionCasted: this should not happen!"
-
-class Typeable thing
-    => UnifiedThing unifiedThing thing | unifiedThing -> thing
-  where
-    destructor :: unifiedThing -> Either (thing Meta) (thing Object)
-    objectConstructor :: thing Object -> unifiedThing
-    metaConstructor :: thing Meta -> unifiedThing
-    transformUnified
-        :: (forall level . MetaOrObject level => thing level -> b)
-        -> (unifiedThing -> b)
-    transformUnified f unifiedStuff =
-        case destructor unifiedStuff of
-            Left x  -> f x
-            Right x -> f x
-    asUnified :: MetaOrObject level => thing level -> unifiedThing
-    asUnified x = applyMetaObjectFunction x MetaOrObjectTransformer
-        { objectTransformer = objectConstructor
-        , metaTransformer = metaConstructor
-        }
+    isMetaOrObject _ = IsObject
+
+getMetaOrObjectType :: MetaOrObject level => thing level -> IsMetaOrObject level
+getMetaOrObjectType _ = isMetaOrObject (undefined :: level)
+
+data Unified thing
+    = UnifiedObject !(thing Object)
+    | UnifiedMeta !(thing Meta)
+
+type ShowMetaOrObject thing = (Show (thing Meta), Show (thing Object))
+type EqMetaOrObject thing = (Eq (thing Meta), Eq (thing Object))
+type OrdMetaOrObject thing = (Ord (thing Meta), Ord (thing Object))
+
+deriving instance (EqMetaOrObject thing) => Eq (Unified thing)
+deriving instance (OrdMetaOrObject thing) => Ord (Unified thing)
+deriving instance (ShowMetaOrObject thing) => Show (Unified thing)
+
+applyUnified
+    :: (thing Meta -> b)
+    -> (thing Object -> b)
+    -> (Unified thing -> b)
+applyUnified metaT objectT =
+    \case
+        UnifiedMeta x -> metaT x
+        UnifiedObject x -> objectT x
+
+transformUnified
+    :: (forall level . MetaOrObject level => thing level -> b)
+    -> (Unified thing -> b)
+transformUnified f = applyUnified f f
+
+mapUnified
+    :: (forall level . thing1 level -> thing2 level)
```

Comment:
Why should this argument accept only functions which work for all `level`s and why isn't it enough to make it work on `MetaOrObject` levels?